### PR TITLE
Issue #2350: Portfolio domain permissions

### LINKED ISSUE #2350: Portfolio user can view/edit domains based on their permissions
### Issue description

The portfolio user can view domains in the portfolio. The ability to **manage** the domains in a portfolio is restricted. An user can manage a domain (edit domain information) if they are defined as a domain manager of the domain. Otherwise, the user can view the domain information but can not edit the domain information

### Acceptance criteria

- [ ] only domains in the user's portfolio are shown on the Domains page. If the user has a domain that is NOT part of their portfolio, it will NOT show. (this user would need to reach out to us about a missing domain)
- [ ] A portfolio user **can** edit a domain in the portfolio only if they are defined as a domain manager (note the manage icon should appear in this case)
- [ ] admins and read-only admins can see all domains in their org in this table.
- [ ] admins and read-only admins cannot edit a domain UNLESS they are added as a domain manager to a particular domain. They should see the eye icon (view only) on all domains that they don't have a UserDomainRole for. (Note, the portfolio read-only view that shows when the eye icon is clicked will be made in #2379)
- [ ] The same state rules apply that can make a domain view only (eye icon appears on the overview page), such as Deleted and on hold domains still need to show as view only
- [ ] The domains page/ tab is only shown to those with domain management permissions, read-only admin and administrator privileges  all others would get a 404 page not found when going to this page

### Additional context

Note that this heavily relates to #2379 both could be done in the same PR if desired. 

### Links to other issues
- blocked by #2366 
- related to #2367 
- blocks #2379

### PR DESCRIPTION
**Pull request by dave-kennedy-ecs**
## Ticket

Resolves #2350 

## Changes

- Portfolio user can see portfolio domains in domains list, and has appropriate links to either View or Manage, based on their individual permissions (whether they are a manager of the domain) and on the state of the domain
- NOTE : the View does not work yet, will be part of 2379 ticket, for portfolio users who are not managers of a domain 

## Setup

Make sure that organization_feature waffle flag is Yes, a portfolio exists, the portfolio is assigned to your test user, there are domains associated with the portfolio, and your test user has portfolio permissions assigned.

#### Satisfied acceptance criteria and met development standards

- [ ] Met the acceptance criteria, or will meet them in a subsequent PR
- [ ] Created/modified automated tests
- [ ] Added at least 2 developers as PR reviewers (only 1 will need to approve)
- [ ] Messaged on Slack or in standup to notify the team that a PR is ready for review
- [ ] Changes to “how we do things” are documented in READMEs and or onboarding guide
- [ ] If any model was updated to modify/add/delete columns, makemigrations was ran and the associated migrations file has been commited.

#### Ensured code standards are met (Original Developer)

- [ ] All new functions and methods are commented using plain language
- [ ] Did dependency updates in Pipfile also get changed in requirements.txt?
- [ ] Interactions with external systems are wrapped in try/except
- [ ] Error handling exists for unusual or missing values

#### Validated user-facing changes (if applicable)

- [ ] New pages have been added to .pa11yci file so that they will be tested with our automated accessibility testing
- [ ] Checked keyboard navigability
- [ ] Tested general usability, landmarks, page header structure, and links with a screen reader (such as Voiceover or ANDI)
- [ ] Add at least 1 designer as PR reviewer

#### Reviewed, tested, and left feedback about the changes

- [ ] Pulled this branch locally and tested it
- [ ] Reviewed this code and left comments
- [ ] Checked that all code is adequately covered by tests
- [ ] Made it clear which comments need to be addressed before this work is merged
- [ ] If any model was updated to modify/add/delete columns, makemigrations was ran and the associated migrations file has been commited.

#### Ensured code standards are met (Code reviewer)

- [ ] All new functions and methods are commented using plain language
- [ ] Interactions with external systems are wrapped in try/except
- [ ] Error handling exists for unusual or missing values
- [ ] (Rarely needed) Did dependency updates in Pipfile also get changed in requirements.txt?

#### Validated user-facing changes as a developer

- [ ] New pages have been added to .pa11yci file so that they will be tested with our automated accessibility testing
- [ ] Checked keyboard navigability
- [ ] Meets all designs and user flows provided by design/product
- [ ] Tested general usability, landmarks, page header structure, and links with a screen reader (such as Voiceover or ANDI)

- [ ] Tested with multiple browsers, the suggestion is to use ones that the developer didn't (check off which ones were used)
  - [ ] Chrome
  - [ ] Microsoft Edge
  - [ ] FireFox
  - [ ] Safari

- [ ] (Rarely needed) Tested as both an analyst and applicant user

**Note:** Multiple code reviewers can share the checklists above, a second reviewers should not make a duplicate checklist

#### Verified that the changes match the design intention

- [ ] Checked that the design translated visually
- [ ] Checked behavior
- [ ] Checked different states (empty, one, some, error)
- [ ] Checked for landmarks, page heading structure, and links
- [ ] Tried to break the intended flow

#### Validated user-facing changes as a designer

- [ ] Checked keyboard navigability
- [ ] Tested general usability, landmarks, page header structure, and links with a screen reader (such as Voiceover or ANDI)

- [ ] Tested with multiple browsers (check off which ones were used)
  - [ ] Chrome
  - [ ] Microsoft Edge
  - [ ] FireFox
  - [ ] Safari

- [ ] (Rarely needed) Tested as both an analyst and applicant user

### CODE REVIEW

**Comment by erinysong** [comment]
Testing on sandbox and got a 403 error from the following: I created a portfolio and populated it with domains and assigned myself as an admin role to the portfolio with no additional portfolio permissions. I am able to view the domains in a portfolio in the portfolio view but 1. the portfolio domains give the option to View when I am an admin and 2. clicking on a domain in my portfolio leads me to a 403 error. The same happens when I assign myself all portfolio permissions.
image
image

Is there another step in between I should do to set proper permissions to use my assigned portfolio? Also happy to hop on a call to verify if what I did accurately reflects how admin would assign portfolios if that is faster!

**Comment by dave-kennedy-ecs** [comment]
> Testing on sandbox and got a 403 error from the following: I created a portfolio and populated it with domains and assigned myself as an admin role to the portfolio with no additional portfolio permissions. I am able to view the domains in a portfolio in the portfolio view but 1. the portfolio domains give the option to View when I am an admin and 2. clicking on a domain in my portfolio leads me to a 403 error. The same happens when I assign myself all portfolio permissions. image image image
> 
> Is there another step in between I should do to set proper permissions to use my assigned portfolio? Also happy to hop on a call to verify if what I did accurately reflects how admin would assign portfolios if that is faster!

That is actually the expected behavior for this ticket.  In the next ticket, 2379, I am implementing the domain view when a portfolio user, who does not have a UserDomainRole for the domain, clicks View on the domain.

**Comment by abroddrick** [comment]
Hey @dave-kennedy-ecs I assigned myself to the portfolio Erin made and gave myself the admin role and request access. But as an admin I am still not seeing all domains in the org. Switched it to read only admin as well, same thing. Both should be able to see all domains in their org
image

**Comment by dave-kennedy-ecs** [comment]
> Hey @dave-kennedy-ecs I assigned myself to the portfolio Erin made and gave myself the admin role and request access. But as an admin I am still not seeing all domains in the org. Switched it to read only admin as well, same thing. Both should be able to see all domains in their org image

@abroddrick there was a problem with the javascript file, getgov.js, being cached by the server.  I redeployed to the dk sandbox, and it is now using the correct javascript file.  Can you test again?

File: src/registrar/views/domains_json.py [hunk 1]
```diff
@@ -43,6 +44,21 @@ def get_domains_json(request):
...
+    """
+    portfolio = request.GET.get("portfolio")
+    if portfolio:
+        domain_infos = DomainInformation.objects.filter(portfolio=portfolio)
+        return domain_infos.values_list("domain_id", flat=True)
+    else:
+        user_domain_roles = UserDomainRole.objects.filter(user=request.user)
+        return user_domain_roles.values_list("domain_id", flat=True)
+
+
```
[the remark this answers is not in the document]
Comment (dave-kennedy-ecs) [reply src/registrar/views/domains_json.py:61]: @zandercymatics I thought about this initially, but thought to make the json service (which has become more of a search function) more flexible, I decided to make the arguments for the json service very deliberate.

**Comment by erinysong** [comment]
Verified that I can assign a user a portfolio, add domains to a portfolio, and view the domains a portfolio on the registrar page with the correct view/manage permissions
image

**Review by erinysong — APPROVED** [review]
LGTM - verified changes successfully work on sandbox and local. I'm good with this being merged but feel free to keep open the discussion with Zander on how strictly we want to set our portfolio display logic

**Pull request merged**